automap/src/comm_layer/pcp.rs

```diff
@@ -1563,10 +1908,78 @@ mod tests {
         );
     }
 
+    #[test]
+    fn thread_guts_logs_and_continues_if_announcement_is_received_before_mapping_config() {
+        init_test_logging();
+        let (tx, rx) = unbounded();
+        let mut announce_packet = vanilla_response();
+        announce_packet.opcode = Opcode::Announce;
+        announce_packet.lifetime = 0;
+        let mut announce_buf = [0u8; 100];
+        let announce_packet_len = announce_packet.marshal(&mut announce_buf).unwrap();
+        let socket: Box<dyn UdpSocketWrapper> = Box::new(
+            UdpSocketWrapperMock::new()
+                .set_read_timeout_result(Ok(()))
+                .recv_from_result(
+                    Ok((
+                        announce_packet_len,
+                        SocketAddr::from_str("1.1.1.1:1111").unwrap(),
+                    )),
+                    announce_buf.to_vec(),
+                ),
+        );
+        let change_log_arc = Arc::new(Mutex::new(vec![]));
+        let inner_cla = change_log_arc.clone();
+        let change_handler = Box::new(move |change| {
+            let mut change_log = inner_cla.lock().unwrap();
+            change_log.push(change)
+        });
+        let mapping_transactor = Box::new(
+            MappingTransactorMock::new().transact_result(Ok((1111, *vanilla_map_response()))),
+        );
+        let logger = Logger::new(
+            "thread_guts_logs_and_continues_if_announcement_is_received_before_mapping_config",
+        );
+
+        let handle = thread::spawn(move || {
+            let _ = PcpTransactor::thread_guts(
+                socket.as_ref(),
+                &rx,
+                Arc::new(Mutex::new(PcpTransactorInner {
+                    mapping_transactor,
+                    factories: Factories::default(),
+                })),
+                SocketAddr::new(IpAddr::from_str("1.1.1.1").unwrap(), 0),
+                change_handler,
+                10,
+                logger,
+            );
+        });
+
+        thread::sleep(Duration::from_millis(100));
```

Comment:
I don't see how it could work here. All I have in the `spawn` is a single blocking function call. If I send the signal after the call returns, it'll never happen, and if I send it before the call is made, there's no advantage over the code that's sitting there now.